### DISCUSSION TITLE
Issues with whitespace

### DISCUSSION THREAD

**Discussion by fatlazycat** [pointfreeco/swift-parsing#76] [Q&A]
Doing AoC 2021 I had the input 

```
41 31 12 14 82
 4 97 76 49 15
50 43 72 22 24
53 56 78 33 52
65 68 26  0 94
```

How would you parse this ?

If I had 

```
41,31,12,14,82
4,97,76,49,15
50,43,72,22,24
53,56,78,33,52
65,68,26,0,94
```

I would just use for each row

```
Many(Int.parser(), separator: StartsWith(","))
```

But unsure where I would start attempting it where you might have leading whitespace.

I looked to explicitly stating Int.parser() with skipping whitespace but can't seem to grasp it, such as 

```
        let parser = Prefix(while: \.isWhitespace)
            .take(Int.parser())
            .skip(Whitespace())
            .take(Int.parser())
            .skip(Whitespace())
            .take(Int.parser())
            .skip(Whitespace())
            .take(Int.parser())
            .skip(Whitespace())
            .take(Int.parser())
```

Thanks

**Comment by iampatbrown** [comment]
One way to approach it could be

```swift
var input = """
41 31 12 14 82
 4 97 76 49 15
50 43 72 22 24
53 56 78 33 52
65 68 26  0 94
"""[...].utf8

let output = Many(
  Skip(Whitespace())
    .take(Int.parser())
    .skip(Whitespace())
).parse(&input)
```

**Reply by fatlazycat** [reply]
Thanks, if you were to just define the parser without giving the type info from the input what you do ?

I'd like to parse each line individually.

**Comment by fatlazycat** [comment]
Came up with

```
        let dataSet = Many(PrefixUpTo<Substring>("\n"), separator: "\n").parse(data)!
        let parser = Many(Skip(Whitespace<Substring.UTF8View>()).take(Int.parser()).skip(Whitespace()))
        let result = dataSet.map{ parser.parse($0.utf8) }.map { $0.0! }
```

My lack of understanding of Strings in Swift seems the main issue 😀

**Reply by stephencelis** [reply]
Another way of breaking the problem down and staying in `Substring`:

```swift
let spaces = Skip(Prefix(while: { $0 == " " }))
let row = Many(spaces.take(Int.parser()).skip(spaces))
let rows = Many(row, separator: "\n")
```

**Reply by fatlazycat** [reply]
Needed todo

```
        let spaces = Skip(Prefix<Substring>(while: { $0 == " " }))
```

But that looks nice & clean, thanks

**Reply by fatlazycat** [reply]
@stephencelis out of interest is there a reason why Whitespace is only for UTF8 ?

**Reply by stephencelis** [reply]
`Whitespace()` only consumes ASCII whitespace at the moment. It should probably be spelled `ASCII.Whitespace()` in the future, which would open up the opportunity for a `Whitespace` that works with `Substring`.

**Reply by fatlazycat** [reply]
@stephencelis See you have released a new version with the aid for type inference. Thanks
